src/library/attribute_manager.cpp

```diff
@@ -7,121 +7,161 @@ Author: Leonardo de Moura
 #include <vector>
 #include <string>
 #include <algorithm>
+#include "util/priority_queue.h"
 #include "util/sstream.h"
 #include "library/attribute_manager.h"
+#include "library/scoped_ext.h"
 
 namespace lean {
-struct attribute_decl {
-    std::string        m_id;
-    std::string        m_descr;
-    std::string        m_token;
-    has_attribute_proc m_tester;
-};
 
-struct default_attribute_decl : public attribute_decl {
-    set_attribute_proc m_setter;
-};
+static name_map<attribute_ptr> * g_attributes;
```

Comment:
Why is this a name? At least currently, you seem to only be storing strings here. Also, you may want to make it a hash table instead of a functional map.